[JSC] Fix `ToIndex(value)` to Align with TC39 spec

Gumichocopengin8 opened the pull request:
#### a0a071d660a164dc5e4e40ec7bd0219a97f6db19
<pre>
[JSC] Fix `ToIndex(value)` to Align with TC39 spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=298894">https://bugs.webkit.org/show_bug.cgi?id=298894</a>

Reviewed by Darin Adler and Yusuke Suzuki.

This patch fixes `ToIndex(value)` implementation to align with TC39 [1].
This fix indirectly corrects functions that rely on `ToIndex(value)`,
such as
- `BigInt.asUintN` [2]
- `BigInt.asIntN` [3]
- `GetViewValue` [4]
- `SetViewValue` [5]
- `ValidateAtomicAccess` [6]

[1]: <a href="https://tc39.es/ecma262/#sec-toindex">https://tc39.es/ecma262/#sec-toindex</a>
[2]: <a href="https://tc39.es/ecma262/#sec-bigint.asuintn">https://tc39.es/ecma262/#sec-bigint.asuintn</a>
[3]: <a href="https://tc39.es/ecma262/#sec-bigint.asintn">https://tc39.es/ecma262/#sec-bigint.asintn</a>
[4]: <a href="https://tc39.es/ecma262/#sec-getviewvalue">https://tc39.es/ecma262/#sec-getviewvalue</a>
[5]: <a href="https://tc39.es/ecma262/#sec-setviewvalue">https://tc39.es/ecma262/#sec-setviewvalue</a>
[6]: <a href="https://tc39.es/ecma262/#sec-validateatomicaccess">https://tc39.es/ecma262/#sec-validateatomicaccess</a>

* JSTests/stress/SharedArrayBuffer.js:
* JSTests/stress/bigint-asintn.js:
(shouldThrow):
* JSTests/stress/bigint-asuintn.js:
(shouldThrow):
* JSTests/stress/data-view-bigint.js:
(shouldBe):
(shouldThrow):
* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::getWaiterListSize):
* Source/JavaScriptCore/runtime/BigIntConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::toIndex const):
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
(JSC::getData):
(JSC::setData):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
(JSC::JSGenericTypedArrayView&lt;Adaptor&gt;::inBounds const):

Canonical link: <a href="https://commits.webkit.org/300157@main">https://commits.webkit.org/300157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a6d07b698cb9f38a7bb7af8d74504c1f739b5b8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73570 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49763 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92273 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61389 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33441 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72950 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32450 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26986 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71511 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113617 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102927 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27160 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130762 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120007 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48415 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36807 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100860 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105045 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100767 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25557 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46176 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24256 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45111 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48274 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53986 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150169 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47745 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51091 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49427 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -65,27 +65,34 @@ inline uint32_t JSValue::toUInt32(JSGlobalObject* globalObject) const
+            throwException(globalObject, scope, createRangeError(globalObject, makeString(errorName, " cannot be negative"_s)));
+            return 0;
+        }
+        return static_cast<uint64_t>(integer);
```
What compiler error do we get if we write this without the cast?
```suggestion
        return integer;
```

Gumichocopengin8 replied:
no compile error with the change. updated.

Gumichocopengin8 commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -121,10 +121,10 @@ EncodedJSValue atomicReadModifyWriteCase(JSGlobalObject* globalObject, VM& vm, c
 }
 
-static unsigned validateAtomicAccess(JSGlobalObject* globalObject, VM& vm, JSArrayBufferView* typedArrayView, JSValue accessIndexValue)
+static uint64_t validateAtomicAccess(JSGlobalObject* globalObject, VM& vm, JSArrayBufferView* typedArrayView, JSValue accessIndexValue)
```
Not sure if we should return `size_t` here or `uint64_t`. 
FireFox cast `uint64_t` to `size_t`, and returns `size_t`, but I think it could be unsafe.

https://github.com/mozilla-firefox/firefox/blob/938e8f38c6765875e998d5c2965ad5864f5a5ee2/js/src/builtin/AtomicsObject.cpp#L129-L156

Constellation replied:
In 64bit, `size_t` is the same size to `uint64_t`, so basically it is ok. But you can just use `uint64_t` given that this is array index which can be larger than `uint32_t` range.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -572,12 +572,12 @@ EncodedJSValue getWaiterListSize(JSGlobalObject* globalObject, CallFrame* callFr
     switch (typedArrayView->type()) {
     case Int32ArrayType: {
-        auto ptr = jsCast<JSInt32Array*>(typedArrayView)->typedVector() + accessIndex;
+        auto ptr = jsCast<JSInt32Array*>(typedArrayView)->typedVector() + static_cast<size_t>(accessIndex);
```
Why does this cast?

Gumichocopengin8 replied:
I thought the explicit casting would be helpful. Will remove it.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -618,7 +618,7 @@ JSC_DEFINE_HOST_FUNCTION(atomicsFuncNotify, (JSGlobalObject* globalObject, CallF
     switch (typedArrayView->type()) {
     case Int32ArrayType: {
-        int32_t* ptr = jsCast<JSInt32Array*>(typedArrayView)->typedVector() + accessIndex;
+        int32_t* ptr = jsCast<JSInt32Array*>(typedArrayView)->typedVector() + static_cast<size_t>(accessIndex);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -65,27 +65,34 @@ inline uint32_t JSValue::toUInt32(JSGlobalObject* globalObject) const
     }
 
-    RELEASE_AND_RETURN(scope, JSC::toInt32(d));
+    RELEASE_AND_RETURN(scope, static_cast<uint64_t>(d));
```
We may get `NaN` here, and casting NaN to `uint64_t` is not correct (`JSC::toInt32` handles it correctly).

Gumichocopengin8 replied:
`d` would contains the value with `uint64_t` range here. I don't think `JSC::toInt32` is suitable.
Also, `toIntegerOrInfinity` returns negative or positive infinity or finite double value (not returning `NaN`), so I think the current approach is safe.
https://tc39.es/ecma262/#sec-tointegerorinfinity

darinadler replied:
I agree that `toIntegerOrInfinity` never returns NaN and I checked our implementation.

What compiler error do we get if we write this without the cast?
```suggestion
    RELEASE_AND_RETURN(scope, d);
```

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -159,12 +159,15 @@ EncodedJSValue getData(JSGlobalObject* globalObject, CallFrame* callFrame)
 
     size_t byteLength = byteLengthValue.value();
-    if (elementSize > byteLength || byteOffset > byteLength - elementSize)
+    if (byteOffset > std::numeric_limits<size_t>::max()
```
Why do we need this check additionally?

Gumichocopengin8 replied:
In 32bit env, `uint64_t` is too large for `size_t`. That's why the range check is added. Is it unnecessary?

darinadler replied:
Yes, this check is unnecessary and this should be left as is. If `byteOffset` is greater than the maximum `size_t` it will also be greater than `byteLength - elementSize`, and so that check will be true.

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -159,12 +159,15 @@ EncodedJSValue getData(JSGlobalObject* globalObject, CallFrame* callFrame)
     constexpr unsigned dataSize = sizeof(typename Adaptor::Type);
     std::array<uint8_t, dataSize> rawBytes { };
-    uint8_t* dataPtr = static_cast<uint8_t*>(dataView->vector()) + byteOffset;
+    uint8_t* dataPtr = static_cast<uint8_t*>(dataView->vector()) + static_cast<size_t>(byteOffset);
```
Why does this cast?

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -207,10 +210,13 @@ EncodedJSValue setData(JSGlobalObject* globalObject, CallFrame* callFrame)
 
     size_t byteLength = byteLengthValue.value();
-    if (elementSize > byteLength || byteOffset > byteLength - elementSize)
+    if (byteOffset > std::numeric_limits<size_t>::max()
```
Ditto.

darinadler replied:
Yes, this check is unnecessary and this should be left as is. If `byteOffset` is greater than the maximum `size_t` it will also be greater than `byteLength - elementSize`, and so that check will be true.

Constellation commented on Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
```diff
@@ -207,10 +210,13 @@ EncodedJSValue setData(JSGlobalObject* globalObject, CallFrame* callFrame)
         return throwVMRangeError(globalObject, scope, "Out of bounds access"_s);
 
-    uint8_t* dataPtr = static_cast<uint8_t*>(dataView->vector()) + byteOffset;
+    uint8_t* dataPtr = static_cast<uint8_t*>(dataView->vector()) + static_cast<size_t>(byteOffset);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
```diff
@@ -793,10 +793,10 @@ template<typename Adaptor> inline typename Adaptor::Type* JSGenericTypedArrayVie
 {
     if (canUseRawFieldsDirectly()) [[likely]]
-        return i < lengthRaw();
+        return i < static_cast<uint64_t>(lengthRaw());
```
Why do we need this cast? Integer promotion will just extend it, right?

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
```diff
@@ -805,9 +805,9 @@ template<typename Adaptor> inline bool JSGenericTypedArrayView<Adaptor>::inBound
         constexpr size_t logSize = logElementSize(Adaptor::typeValue);
         size_t remainingLength = bufferByteLength - byteOffset;
-        return i < (remainingLength >> logSize);
+        return i < static_cast<uint64_t>(remainingLength >> logSize);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
```diff
@@ -805,9 +805,9 @@ template<typename Adaptor> inline bool JSGenericTypedArrayView<Adaptor>::inBound
+        return i < static_cast<uint64_t>(remainingLength >> logSize);
     }
-    return i < lengthRaw();
+    return i < static_cast<uint64_t>(lengthRaw());
```
Ditto.

darinadler reviewed (approved):
I’m surprised that our `JSValue::toIndex` function is used so little, and also that the changes to places that we are now using `uint64_t` aside from `BigInt` lead to no improvements in test results. You’d think there would be some observable change!

darinadler reviewed (approved):
We claim that this helps functions other than `BigInt.asUintN` and `BigInt.asIntN`. Which other functions? I think that claim is wrong and this change only helps those two functions, and has no other observable effect.

Gumichocopengin8 committed:
[JSC] Fix `ToIndex(value)` to Align with TC39 spec
https://bugs.webkit.org/show_bug.cgi?id=298894

Reviewed by Darin Adler and Yusuke Suzuki.

This patch fixes `ToIndex(value)` implementation to align with TC39 [1].
This fix indirectly corrects functions that rely on `ToIndex(value)`,
such as
- `BigInt.asUintN` [2]
- `BigInt.asIntN` [3]
- `GetViewValue` [4]
- `SetViewValue` [5]
- `ValidateAtomicAccess` [6]

[1]: https://tc39.es/ecma262/#sec-toindex
[2]: https://tc39.es/ecma262/#sec-bigint.asuintn
[3]: https://tc39.es/ecma262/#sec-bigint.asintn
[4]: https://tc39.es/ecma262/#sec-getviewvalue
[5]: https://tc39.es/ecma262/#sec-setviewvalue
[6]: https://tc39.es/ecma262/#sec-validateatomicaccess

* JSTests/stress/SharedArrayBuffer.js:
* JSTests/stress/bigint-asintn.js:
(shouldThrow):
* JSTests/stress/bigint-asuintn.js:
(shouldThrow):
* JSTests/stress/data-view-bigint.js:
(shouldBe):
(shouldThrow):
* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::getWaiterListSize):
* Source/JavaScriptCore/runtime/BigIntConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::toIndex const):
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
(JSC::getData):
(JSC::setData):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
(JSC::JSGenericTypedArrayView<Adaptor>::inBounds const):

Canonical link: https://commits.webkit.org/300157@main
